### PR TITLE
EPHEH-313: Implement Color Scheme

### PR DESCRIPTION
**Pull request by tibi2303**

### CODE REVIEW

File: src/themes/default/src/scss/color_schema.scss [hunk 1]
```diff
@@ -0,0 +1,126 @@
+// Configuration
+@import "@openeuropa/bcl-bootstrap/scss/functions";
```
Comment (drishu) [inline src/themes/default/src/scss/color_schema.scss:2]: the default theme package depends on the bcl bootstrap, can we do something to reflect this in the dependencies section of its package.json? would it break any existing setup?

**Review by drishu — CHANGES_REQUESTED** [review]
- we need some explanation/documentation on how to use it in a project
- also the way sass is handled by toolkit is through the toolkit:build-assets command which can be set up to use ecl-builder or gulp, after leaving instructions on how to go around this we need to follow up on adding a bcl-builder to it

File: builder/README.md [hunk 2]
```diff
@@ -40,3 +40,78 @@ commands which all supports multiple operations as items in an array.
...
+```javascript
+color-scheme: [
+  {
+    entry (ex: ../path/to/your-colour-scheme.scss),
+    dest (ex: ../path/to/compiled.css),
+    options (object with plugins for postcss) {
+      minify (includes css nano for minification)
+    }
+  },
+],
```
Comment (drishu) [inline builder/README.md:57]: Trying to integrate it into oe_showcase I found that this part here is useless

File: builder/README.md [hunk 3]
```diff
@@ -40,3 +40,78 @@ commands which all supports multiple operations as items in an array.
...
+
+ * Example config object: {
+ color-scheme: [
+  {
+    entry: [path.resolve(nodeModules, "color-scheme.scss")],
```
Comment (drishu) [inline builder/README.md:62]:
```suggestion
    entry: path.resolve(outputFolder, "resources/sass/color-scheme-variables.scss"),
```

File: builder/README.md [hunk 4]
```diff
@@ -40,3 +40,78 @@ commands which all supports multiple operations as items in an array.
...
+    dest: path.resolve(outputFolder, "css"),
```
Comment (drishu) [inline builder/README.md:63]:
```suggestion
    dest: path.resolve(outputFolder, "assets/css/color_scheme.min.css"),
```

File: builder/README.md [hunk 5]
```diff
@@ -40,3 +40,78 @@ commands which all supports multiple operations as items in an array.
...
+    entry (ex: ../path/to/your-colour-scheme.scss),
+    dest (ex: ../path/to/compiled.css),
+    options (object with plugins for postcss) {
+      minify (includes css nano for minification)
+    }
+  },
+],
+
+ * Example config object: {
+ color-scheme: [
```
Comment (drishu) [inline builder/README.md:60]:
```suggestion
 colorScheme: [
```

File: builder/README.md [hunk 6]
```diff
@@ -40,3 +40,78 @@ commands which all supports multiple operations as items in an array.
...
+    "primary": #4eac00,
+    "secondary": #fbff00,
+    "danger": #ff8800,
+    "success": #4dff00,
+    "text": #294d26,
+    "background": #ffdfb0,
+    "link": #7b00ff,
+  ),
+);
+```
```
Comment (drishu) [inline builder/README.md:86]: Lets explain that this is the contents of the color-scheme-variables.scss

File: builder/README.md [hunk 7]
```diff
@@ -40,3 +40,78 @@ commands which all supports multiple operations as items in an array.
...
+
+You can find multiple plugins here:
+[Postcss Plugins](https://github.com/postcss/postcss/blob/main/docs/plugins.md)
+
+The color-scheme enables the possibility to theme the colors used in BCL. It overrides the variables `$theme-colors` used by Bootstrap ([Bootstrap Theme Colors](https://getbootstrap.com/docs/5.0/customize/color/#theme-colors)).
+
+### Prerequisites
+
+- BCL-Builder version >= 1.2.1
```
Comment (drishu) [inline builder/README.md:95]: +BCL Bootstrap

But rather let's explain rather what needs to be added into the package, with a minimal setup example (maybe they dont have one yet in their project)

File: builder/README.md [hunk 8]
```diff
@@ -40,3 +40,78 @@ commands which all supports multiple operations as items in an array.
...
+- `scss/color_scheme/background`: Mixin that adds the `bg-default` class
+- `scss/color_scheme/badge`: Mixin that changes the colors of badges
+- `scss/color_scheme/buttons`: Mixin that changes the colors of buttons
+- `scss/color_scheme/link`: Mixin that changes the color of links
+- `scss/color_scheme/list_group`: Mixin that changes the colors of list groups
+- `scss/color_scheme/text`: Mixin that adds the `text-color-default` class
+- `scss/color_scheme/utilities`: Mixin that changes the utilities: `bg-{color}` (e.g., `bg-primary`), `text-bg-{color}` (e.g., `text-bg-primary`), and `link-{color}` (e.g., `link-primary`)
+- `scss/color_scheme/generator`: This includes all the mentioned mixins and calls them. It also adds the CSS variables on the color scheme class.
+
+- `scss/color-scheme.scss`: This is the starting imports needed.
```
Comment (drishu) [inline builder/README.md:115]: useless, remove

Rather mention all that is needed, the package.json, the builder config, variables sass file, node to compile, what commands to run.

File: builder/scripts/colorScheme.js [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+      ...(options.includePaths || []),
+      "@openeuropa/bcl-theme-default/src/scss/color_scheme/generator.scss"
+    ),
+    "utf8"
+  );
+
+  // Concatenate the contents
+  const scssContent = imports + "\n" + entryVariables + "\n" + generator;
+
+  console.log(scssContent);
```
Comment (drishu) [inline builder/scripts/colorScheme.js:72]: leftover, please remove

**Review by drishu — CHANGES_REQUESTED** [review]

File: builder/README.md (hunk 6 above)
Comment (tibi2303) [reply builder/README.md]: updated.

File: builder/README.md (hunk 5 above)
Comment (tibi2303) [reply builder/README.md]: updated.

File: builder/README.md (hunk 4 above)
Comment (tibi2303) [reply builder/README.md]: updated.

File: builder/README.md (hunk 2 above)
Comment (tibi2303) [reply builder/README.md]: removed.

File: builder/README.md (hunk 3 above)
Comment (tibi2303) [reply builder/README.md]: updated.

File: builder/README.md (hunk 7 above)
Comment (tibi2303) [reply builder/README.md:91]: updated.

File: builder/scripts/colorScheme.js (hunk 9 above)
Comment (tibi2303) [reply builder/scripts/colorScheme.js]: removed.

File: builder/README.md (hunk 8 above)
Comment (tibi2303) [reply builder/README.md]: Overall i've changed the readme and also provided a basic setup for color schema.

File: builder/README.md (hunk 3 above)
Comment (drishu) [reply builder/README.md]: not quite, the `"resources/sass/color-scheme-variables.scss"` part is missing

File: builder/scripts/colorScheme.js [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+ *  color-scheme: [
+ *   {
+ *     entry: [path.resolve(nodeModules, "myfile.js")],
+ *     dest: path.resolve(outputFolder, "js"),
+ *     options: {
+ *       includePaths,
+ *       sourceMap: "file",
+ *     },
+ *   },
+ * ],
```
Comment (drishu) [inline builder/scripts/colorScheme.js:19]: you should have updated alo this or remove

**Review by drishu — CHANGES_REQUESTED** [review]
Working on the showcase intregration I noticed:

- Carousel does not recipe any color changes
- Gallery receives no color changes
- Timeline circles stay blue

Also one point remains from previous review

File: builder/scripts/colorScheme.js (hunk 10 above)
Comment (tibi2303) [reply builder/scripts/colorScheme.js]: I've updated.

Also i fixed the timeline circle and the image banner bottom. The carousel and gallery from my point of view are behaving correctly. I added 2 stories for each of them inside color schema.

**Review by drishu — APPROVED** [review]

File: src/themes/default/src/scss/color_scheme/_utilities.scss [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+@mixin generate-utilities($color, $value) {
+  $color-rgb: to-rgb($value);
+  .text-bg-#{$color} {
+    color: color-contrast($value)
+      if($enable-important-utilities, !important, null);
+    background-color: RGBA($color-rgb, var(--#{$prefix}bg-opacity, 1))
+      if($enable-important-utilities, !important, null);
+  }
+
+  .bg-#{$color} {
```
Comment (tibi2303) [inline src/themes/default/src/scss/color_scheme/_utilities.scss:12]: we need also to add text-bg-{#color} utilities

**Review by drishu — APPROVED** [review]

**Review by brummbar — APPROVED** [review]

**Review by drishu — APPROVED** [review]

**Review by brummbar — APPROVED** [review]

**Pull request merged**